File: livingwell-frontend/src/types/lifecycle.ts
// ── Property Lifecycle Types ─────────────────────────────────────────

export type DevelopmentStage =
  | "prospect"
  | "acquisition"
  | "interim_operation"
  | "planning"
  | "permit"
  | "construction"
  | "lease_up"
  | "stabilized"
  | "exit";

export type MilestoneStatus = "pending" | "in_progress" | "completed" | "skipped";

export interface StageTransition {
  transition_id: number;
  property_id: number;
  from_stage: DevelopmentStage;
  to_stage: DevelopmentStage;
  transitioned_by: number;
  transitioned_at: string;
  notes: string | null;
  validation_passed: boolean;
  validation_errors: string | null;
}

export interface AllowedTransition {
  current_stage: DevelopmentStage;
  allowed_transitions: DevelopmentStage[];
  validation_requirements?: Record<string, string[]>;
}

export interface TransitionRequest {
  to_stage: DevelopmentStage;
  notes?: string;
  force?: boolean;
}

export interface PropertyMilestone {
  milestone_id: number;
  property_id: number;
  title: string;
  description: string | null;
  status: MilestoneStatus;
  stage: DevelopmentStage;
  target_date: string;
  actual_date: string | null;
  sort_order: number;
}

export interface MilestoneCreate {
  title: string;
  description?: string;
  target_date: string;
  stage: DevelopmentStage;
  sort_order?: number;
}

export interface MilestoneUpdate {
  title?: string;
  description?: string;
  status?: MilestoneStatus;
  target_date?: string;
  actual_date?: string;
}

// ── Quarterly Reports ───────────────────────────────────────────────

export type QuarterlyReportStatus = "draft" | "review" | "published";

export interface QuarterlyReport {
  report_id: number;
  lp_id: number;
  period_label: string;
  quarter: number;
  year: number;
  status: QuarterlyReportStatus;
  total_revenue: string;
  total_expenses: string;
  net_operating_income: string;
  total_distributions: string;
  portfolio_value: string;
  portfolio_ltv: string;
  executive_summary: string | null;
  property_updates: string | null;
  market_commentary: string | null;
  generated_at: string | null;
  published_at: string | null;
  generated_by: number | null;
}

export interface QuarterlyReportCreate {
  quarter: number;
  year: number;
}

export interface QuarterlyReportUpdate {
  status?: QuarterlyReportStatus;
  executive_summary?: string;
  market_commentary?: string;
}

// ── eTransfer Tracking ──────────────────────────────────────────────

export type ETransferStatus =
  | "initiated"
  | "sent"
  | "accepted"
  | "declined"
  | "expired"
  | "cancelled";

export interface ETransferTracking {
  tracking_id: number;
  allocation_id: number;
  recipient_email: string;
  amount: string;
  security_question: string | null;
  reference_number: string | null;
  status: ETransferStatus;
  initiated_at: string | null;
  sent_at: string | null;
  accepted_at: string | null;
  expires_at: string | null;
  notes: string | null;
}

export interface ETransferCreate {
  allocation_id: number;
  recipient_email: string;
  amount: number;
  security_question?: string;
}

export interface ETransferUpdate {
  status?: ETransferStatus;
  reference_number?: string;
  notes?: string;
}

// ── Message Threads ─────────────────────────────────────────────────

export interface MessageReply {
  reply_id: number;
  parent_message_id: number;
  sender_id: number;
  body: string;
  sent_at: string;
  is_read: boolean;
}

export interface MessageReplyCreate {
  body: string;
}

// ── Operator Budgets ────────────────────────────────────────────────

export type BudgetPeriodType = "monthly" | "quarterly" | "annual";

export interface OperatorBudget {
  budget_id: number;
  operator_id: number;
  community_id: number;
  period_type: BudgetPeriodType;
  period_label: string;
  year: number;
  quarter: number | null;
  month: number | null;
  budgeted_revenue: string;
  budgeted_expenses: string;
  budgeted_noi: string;
  actual_revenue: string | null;
  actual_expenses: string | null;
  actual_noi: string | null;
  notes: string | null;
}

export interface BudgetCreate {
  operator_id: number;
  community_id: number;
  period_type: BudgetPeriodType;
  period_label: string;
  year: number;
  quarter?: number;
  month?: number;
  budgeted_revenue: number;
  budgeted_expenses: number;
  budgeted_noi: number;
}

export interface BudgetUpdate {
  actual_revenue?: number;
  actual_expenses?: number;
  actual_noi?: number;
  notes?: string;
}

// ── Operating Expenses ──────────────────────────────────────────────

export type ExpenseCategory =
  | "property_management"
  | "utilities"
  | "insurance"
  | "property_tax"
  | "maintenance_repairs"
  | "staffing"
  | "meal_program"
  | "supplies"
  | "marketing"
  | "technology"
  | "professional_fees"
  | "other";

export interface OperatingExpense {
  expense_id: number;
  community_id: number;
  budget_id: number | null;
  category: ExpenseCategory;
  description: string;
  amount: string;
  expense_date: string;
  period_month: number;
  period_year: number;
  vendor: string | null;
  invoice_ref: string | null;
  is_recurring: boolean;
  notes: string | null;
}

export interface ExpenseCreate {
  community_id: number;
  budget_id?: number;
  category: ExpenseCategory;
  description: string;
  amount: number;
  expense_date: string;
  period_month: number;
  period_year: number;
  vendor?: string;
  invoice_ref?: string;
  is_recurring?: boolean;
}

export interface ExpenseUpdate {
  amount?: number;
  description?: string;
  notes?: string;
}

export interface ExpenseSummary {
  community_id: number;
  year: number;
  quarter: number | null;
  total_expenses: string;
  by_category: Record<string, string>;
  expense_count: number;
}

File: backend/app/schemas/lifecycle.py
"""
Pydantic schemas for Property Lifecycle, Quarterly Reports, eTransfer Tracking,
Message Threads, and Operator Budget/Expense management (Phase 3).
"""
import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel

from app.db.models import (
    DevelopmentStage, MilestoneStatus, QuarterlyReportStatus,
    ETransferStatus, ExpenseCategory, ExpensePhase, BudgetPeriodType,
)


# ---------------------------------------------------------------------------
# Property Lifecycle — Stage Transitions
# ---------------------------------------------------------------------------

class StageTransitionRequest(BaseModel):
    to_stage: DevelopmentStage
    notes: Optional[str] = None
    force: bool = False  # GP_ADMIN only


class ValidationCheckOut(BaseModel):
    name: str
    passed: bool
    message: str


class StageTransitionOut(BaseModel):
    transition_id: int
    property_id: int
    from_stage: str
    to_stage: str
    transitioned_by: int
    transitioned_at: datetime.datetime
    notes: Optional[str]
    validation_passed: bool
    validation_checks: List[ValidationCheckOut] = []

    model_config = {"from_attributes": True}


class AllowedTransitionsOut(BaseModel):
    current_stage: str
    allowed_transitions: List[str]


# ---------------------------------------------------------------------------
# Property Milestones
# ---------------------------------------------------------------------------

class MilestoneCreate(BaseModel):
    title: str
    description: Optional[str] = None
    target_date: Optional[datetime.date] = None
    stage: Optional[DevelopmentStage] = None
    sort_order: int = 0


class MilestoneUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    target_date: Optional[datetime.date] = None
    actual_date: Optional[datetime.date] = None
    status: Optional[MilestoneStatus] = None
    sort_order: Optional[int] = None


class MilestoneOut(BaseModel):
    milestone_id: int
    property_id: int
    title: str
    description: Optional[str]
    target_date: Optional[datetime.date]
    actual_date: Optional[datetime.date]
    status: MilestoneStatus
    stage: Optional[str]
    sort_order: int
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Quarterly Reports
# ---------------------------------------------------------------------------

class QuarterlyReportGenerate(BaseModel):
    quarter: int  # 1-4
    year: int


class QuarterlyReportUpdate(BaseModel):
    status: Optional[QuarterlyReportStatus] = None
    executive_summary: Optional[str] = None
    market_commentary: Optional[str] = None


class PropertyUpdateItem(BaseModel):
    property_id: int
    address: str
    city: str
    stage: str
    total_beds: int
    occupied_beds: int
    occupancy_percent: float
    communities: List[str]


class QuarterlyReportOut(BaseModel):
    report_id: int
    lp_id: int
    period_label: str
    quarter: int
    year: int
    status: QuarterlyReportStatus
    total_revenue: Optional[Decimal]
    total_expenses: Optional[Decimal]
    net_operating_income: Optional[Decimal]
    total_distributions: Optional[Decimal]
    portfolio_value: Optional[Decimal]
    portfolio_ltv: Optional[Decimal]
    executive_summary: Optional[str]
    property_updates: Optional[str]  # JSON string
    market_commentary: Optional[str]
    generated_at: datetime.datetime
    published_at: Optional[datetime.datetime]
    generated_by: Optional[int]

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# eTransfer Tracking
# ---------------------------------------------------------------------------

class ETransferCreate(BaseModel):
    allocation_id: int
    recipient_email: str
    amount: Decimal
    security_question: Optional[str] = None
    notes: Optional[str] = None


class ETransferUpdate(BaseModel):
    status: Optional[ETransferStatus] = None
    reference_number: Optional[str] = None
    sent_at: Optional[datetime.datetime] = None
    accepted_at: Optional[datetime.datetime] = None
    notes: Optional[str] = None


class ETransferOut(BaseModel):
    tracking_id: int
    allocation_id: int
    recipient_email: str
    amount: Decimal
    security_question: Optional[str]
    reference_number: Optional[str]
    status: ETransferStatus
    initiated_at: datetime.datetime
    sent_at: Optional[datetime.datetime]
    accepted_at: Optional[datetime.datetime]
    expires_at: Optional[datetime.datetime]
    notes: Optional[str]

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Message Threads (replies)
# ---------------------------------------------------------------------------

class MessageReplyCreate(BaseModel):
    body: str


class MessageReplyOut(BaseModel):
    reply_id: int
    parent_message_id: int
    sender_id: int
    body: str
    sent_at: datetime.datetime
    is_read: bool

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Operator Budgets
# ---------------------------------------------------------------------------

class OperatorBudgetCreate(BaseModel):
    operator_id: int
    community_id: int
    period_type: BudgetPeriodType = BudgetPeriodType.annual
    period_label: str
    year: int
    quarter: Optional[int] = None
    budgeted_revenue: Decimal = Decimal("0")
    budgeted_expenses: Decimal = Decimal("0")
    budgeted_noi: Decimal = Decimal("0")
    notes: Optional[str] = None


class OperatorBudgetUpdate(BaseModel):
    budgeted_revenue: Optional[Decimal] = None
    budgeted_expenses: Optional[Decimal] = None
    budgeted_noi: Optional[Decimal] = None
    actual_revenue: Optional[Decimal] = None
    actual_expenses: Optional[Decimal] = None
    actual_noi: Optional[Decimal] = None
    notes: Optional[str] = None


class OperatorBudgetOut(BaseModel):
    budget_id: int
    operator_id: int
    community_id: int
    period_type: BudgetPeriodType
    period_label: str
    year: int
    quarter: Optional[int]
    budgeted_revenue: Decimal
    budgeted_expenses: Decimal
    budgeted_noi: Decimal
    actual_revenue: Optional[Decimal]
    actual_expenses: Optional[Decimal]
    actual_noi: Optional[Decimal]
    notes: Optional[str]
    created_at: datetime.datetime

    # Computed fields
    revenue_variance: Optional[Decimal] = None
    expense_variance: Optional[Decimal] = None
    noi_variance: Optional[Decimal] = None

    model_config = {"from_attributes": True}


# ---------------------------------------------------------------------------
# Operating Expenses
# ---------------------------------------------------------------------------

class OperatingExpenseCreate(BaseModel):
    community_id: int
    budget_id: Optional[int] = None
    category: ExpenseCategory
    description: str
    amount: Decimal
    expense_date: datetime.date
    period_month: int
    period_year: int
    vendor: Optional[str] = None
    invoice_ref: Optional[str] = None
    is_recurring: bool = False
    phase: Optional[ExpensePhase] = None
    notes: Optional[str] = None


class OperatingExpenseUpdate(BaseModel):
    category: Optional[ExpenseCategory] = None
    description: Optional[str] = None
    amount: Optional[Decimal] = None
    expense_date: Optional[datetime.date] = None
    vendor: Optional[str] = None
    invoice_ref: Optional[str] = None
    is_recurring: Optional[bool] = None
    phase: Optional[ExpensePhase] = None
    notes: Optional[str] = None


class OperatingExpenseOut(BaseModel):
    expense_id: int
    community_id: int
    budget_id: Optional[int]
    category: ExpenseCategory
    description: str
    amount: Decimal
    expense_date: datetime.date
    period_month: int
    period_year: int
    vendor: Optional[str]
    invoice_ref: Optional[str]
    is_recurring: bool
    phase: Optional[str] = None
    notes: Optional[str]
    created_at: datetime.datetime

    model_config = {"from_attributes": True}


class ExpenseSummaryOut(BaseModel):
    """Aggregated expense summary by category for a community/period."""
    community_id: int
    period_label: str
    total_expenses: Decimal
    by_category: dict  # {category: amount}
    expense_count: int
